### PR TITLE
Standardize logging

### PR DESCRIPTION
**Pull request by JGcarv**
I just restarted this because the whole folder structure changed, so I decided it was better to do it now than to rebase everything else later.

I think this time is going a bit better, I managed to create just one instance of `logger` in the main.go and then pass it to functions.

The only thing I'm worried about now, that's why I stopped changing at this point, is that Data Servers uses channels, which believe is some mechanism to run parallel stuff, and I don't know if I need to do some additional work to make the logger function properly. previously it was passed inside an ops structure, and a new instance every time, so I'm not sure if what I'm doing will be fine!

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
> The only thing I'm worried about now, that's why I stopped changing at this point, is that Data Servers uses channels, which believe is some mechanism to run parallel stuff, and I don't know if I need to do some additional work to make the logger function properly. previously it was passed inside an ops structure, and a new instance every time, so I'm not sure if what I'm doing will be fine!

yes that is fine there is nothing special with the logger with working with channels, just pass it around as usual and that is it.

File: cmd/tellor/main.go [hunk 1]
```diff
@@ -106,20 +116,22 @@ func AddDBToCtx(remote bool) error {
 	os.RemoveAll(cfg.DBFile)
 	DB, err := db.Open(cfg.DBFile)
 	if err != nil {
-		return err
+		return fmt.Errorf("Error opening DB: %v\n", err)
```
Comment (krasi-georgiev) [inline cmd/tellor/main.go:119]: using errors.Wrap is better here:

```suggestion
		return errors.Wrap(err,"opening the DB")
```

File: cmd/tellor/main.go [hunk 2]
```diff
@@ -233,135 +250,145 @@ func newDisputeCmd(cmd *cli.Cmd) {
...
+
+					var err error
+					ds, err = ops.CreateDataServerOps(ctx, logger, ch)
+					if err != nil {
+						level.Error(logger).Log("msg", "error creating data server", "err", err)
+						os.Exit(1)
+					}
+					// Start and wait for it to be ready.
+					if err := ds.Start(ctx, logger); err != nil {
+						level.Error(logger).Log("msg", "error starting data server", "err", err)
```
Comment (krasi-georgiev) [inline cmd/tellor/main.go:278]: I failed start should also exit here no?

```suggestion
						level.Error(logger).Log("msg", "error starting data server", "err", err)
						os.Exit(1)
```

File: cmd/tellor/main.go [hunk 3]
```diff
@@ -233,135 +250,145 @@ func newDisputeCmd(cmd *cli.Cmd) {
...
-
-			if ds != nil {
-				dsStopped = !ds.Running
-			} else {
-				dsStopped = true
+			// Start miner
+			DB := ctx.Value(tellorCommon.DataProxyKey).(db.DataServerProxy)
+			v, err := DB.Get(db.DisputeStatusKey)
+			if err != nil {
+				level.Warn(logger).Log("info", "could not get dispute status. Check if staked")
```
Comment (krasi-georgiev) [inline cmd/tellor/main.go:287]:
```suggestion
				level.Warn(logger).Log("msg", "could not get dispute status. Check if staked")
```

lets keep the format the same everywhere. See prometheus and thanos for examples.
when you have `level.Info` it automatically adds INFO to the logs. the same for warn and error logs

File: cmd/tellor/main.go [hunk 4]
```diff
@@ -233,135 +250,145 @@ func newDisputeCmd(cmd *cli.Cmd) {
...
+
+				if !dsStopped && !minerStopped && cnt > 60 {
+					level.Warn(logger).Log("msg", "Taking longer than expected to stop operations", "waited", time.Since(start))
+				} else if dsStopped && minerStopped {
+					break
+				}
+				time.Sleep(500 * time.Millisecond)
 			}
-			time.Sleep(500 * time.Millisecond)
+			level.Info(logger).Log("info", "Main shutdown complete")
```
Comment (krasi-georgiev) [inline cmd/tellor/main.go:335]:
```suggestion
			level.Info(logger).Log("msg", "Main shutdown complete")
```

File: cmd/tellor/main.go (hunk 1 above)
Comment (JGcarv) [reply cmd/tellor/main.go:119]: I'll leave this as is right now because this same thing is used all over the code. But if it's worth the time, I can do another PR to adjust those all  in one batch. What do you think?

**Comment by JGcarv** [comment]
I don't understand how to make the linting not fail... It's complaining about not checking for errors in the Log function!

File: cmd/tellor/main.go (hunk 1 above)
Comment (krasi-georgiev) [reply cmd/tellor/main.go:119]: I think it is fine to update gradually, since you are changing the code here anyway why not just use the proper form and the rest can do in follow up PRs.

**Comment by krasi-georgiev** [comment]
> I don't understand how to make the linting not fail... It's complaining about not checking for errors in the Log function!

it is fine to exclude this check. See https://github.com/krasi-georgiev/prometheus/blob/45506841e664665e8f0b1b59f416c91643913a3f/scripts/errcheck_excludes.txt#L10

File: pkg/tracker/currentVariables_test.go [hunk 5]
```diff
@@ -29,8 +30,23 @@ func TestCurrentVariables(t *testing.T) {
 
 	tracker := &CurrentVariablesTracker{}
 
+	ctx := context.WithValue(context.Background(), tellorCommon.ClientContextKey, client)
```
Comment (krasi-georgiev) [inline pkg/tracker/currentVariables_test.go:33]: This should be all done in the `testutil.CreateContext(t)`, no? If it isn't best to add it there.

File: pkg/tracker/disputeChecker.go [hunk 6]
```diff
@@ -147,11 +149,22 @@ func (c *disputeChecker) Exec(ctx context.Context) error {
...
-			if result == nil {
-				disputeLogger.Warn("no value data for reqID %d at %s", reqID, blockTime)
-				continue
+		reqID := nonceSubmit.RequestId.Uint64()
+		result := CheckValueAtTime(reqID, nonceSubmit.Value, blockTime)
+		if result == nil {
+			level.Warn(logger).Log("msg", "no value data", "reqid", reqID, "blockTime", blockTime)
+			continue
+		}
+		if !result.WithinRange {
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeChecker.go:158]: this seems to be a duplicate of the one below.

File: pkg/tracker/disputeChecker_test.go [hunk 7]
```diff
@@ -43,8 +45,9 @@ func TestDisputeCheckerOutOfRange(t *testing.T) {
 	execEthUsdPsrs(ctx, t, ethUSDPairs)
 	time.Sleep(2 * time.Second)
 	execEthUsdPsrs(ctx, t, ethUSDPairs)
+	ctx = context.WithValue(ctx, tellorCommon.ContractAddress, common.Address{0x0000000000000000000000000000000000000000})
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeChecker_test.go:48]: I added this in `testutil.CreateContext(t)` so shouldn't be needed

File: pkg/tracker/disputeStatus_test.go [hunk 8]
```diff
@@ -37,7 +39,7 @@ func TestDisputeStatus(t *testing.T) {
 	tracker := &DisputeTracker{}
 	ctx := context.WithValue(context.Background(), common.ClientContextKey, client)
 	ctx = context.WithValue(ctx, common.DBContextKey, DB)
-	err = tracker.Exec(ctx)
+	err = tracker.Exec(ctx, logger)
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeStatus_test.go:42]: for testing where we don't need the logs go kit also offers noop logger

**Comment by krasi-georgiev** [comment]
at the end can also remove the tellor log util package  `pkg/util/logger.go`

File: pkg/tracker/disputeChecker_test.go (hunk 7 above)
Comment (JGcarv) [reply pkg/tracker/disputeChecker_test.go:50]: Ok! I'll remove it.

File: pkg/ops/disputeOps.go [hunk 9]
```diff
@@ -171,7 +172,7 @@ func getNonceSubmissions(ctx context.Context, valueBlock *big.Int, dispute *tell
 	return timedValues, nil
 }
 
-func List(ctx context.Context) error {
+func List(ctx context.Context, logger log.Logger) error {
```
Comment (krasi-georgiev) [inline pkg/ops/disputeOps.go:175]:
```suggestion
func List(ctx context.Context) error {
```

as discussed it looks like we don't need a logger here

File: pkg/tracker/balance.go [hunk 10]
```diff
@@ -43,10 +43,12 @@ func (b *BalanceTracker) Exec(ctx context.Context) error {
...
 
 	if err != nil {
-		fmt.Println("balance Error, balance.go")
+		level.Error(logger).Log("msg", "error getting balance", "err", err)
 		return err
 	}
 	enc := hexutil.EncodeBig(balance)
-	log.Printf("Balance: %v", enc)
+
+	// log.Printf("Balance: %v", enc)
```
Comment (krasi-georgiev) [inline pkg/tracker/balance.go:51]:
```suggestion
```

File: pkg/testutil/testutil.go [hunk 11]
```diff
@@ -55,21 +57,30 @@ func CreateContext(t *testing.T) (context.Context, *config.Config, func()) {
...
 	if err != nil {
-		log.Fatal(err)
+		t.Fatalf("Problem creating proxy: %v\n", err)
 	}
 	ctx = context.WithValue(ctx, common.DataProxyKey, proxy)
 
 	return ctx, cfg, cleanup
 }
+
+func SetupLogger() log.Logger {
```
Comment (krasi-georgiev) [inline pkg/testutil/testutil.go:79]: instead of duplicating the func why not put it in `pkg/util/logger.go` and then can use for the tests and in the main entry point.

File: pkg/tracker/disputeStatus.go [hunk 12]
```diff
@@ -78,13 +79,14 @@ func (b *DisputeTracker) Exec(ctx context.Context) error {
 		//fmt.Println("Getting staker info for address", addr)
 		status, _, err := instance.GetStakerInfo(nil, address)
 		if err != nil {
-			fmt.Printf("Could not get staker dispute status for miner address %s: %v\n", addr, err)
+			level.Error(logger).Log("msg", "Could not get staker dispute status for miner", "address", addr, "err", err)
 		}
 		fmt.Printf("Whitelisted Miner %s Dispute Status: %v\n", addr, status)
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeStatus.go:84]:
```suggestion
```

File: pkg/tracker/runner.go [hunk 13]
```diff
@@ -31,15 +31,17 @@ func NewRunner(client rpc.ETHClient, db db.DB) (*Runner, error) {
...
+func (r *Runner) Start(ctx context.Context, logger log.Logger, exitCh chan int) error {
 	cfg := config.GetConfig()
 	trackerNames := cfg.Trackers
 	var trackers []Tracker
 	for name, activated := range trackerNames {
 		if activated {
 			t, err := createTracker(name)
 			if err != nil {
-				runnerLog.Error("Problem creating tracker: %s\n", err.Error())
+				level.Warn(logger).Log("msg", "Problem creating tracker", "name", name, "err", err)
```
Comment (krasi-georgiev) [inline pkg/tracker/runner.go:42]: I don't think it makes sense to continue if one of the trackers fails to create.
Not everyone looks at the logs to I think this should propagate the error and the caller should decide what to to.

File: pkg/tracker/runner.go [hunk 14]
```diff
@@ -53,21 +55,22 @@ func (r *Runner) Start(ctx context.Context, exitCh chan int) error {
...
+	level.Info(logger).Log("msg", fmt.Sprintf("Created %d trackers", len(trackers)))
 
 	var err error
 	masterInstance := ctx.Value(tellorCommon.ContractsTellorContextKey)
 	if masterInstance == nil {
 		contractAddress := common.HexToAddress(cfg.ContractAddress)
 		masterInstance, err = tellor.NewTellor(contractAddress, r.client)
 		if err != nil {
-			runnerLog.Error("Problem creating tellor master instance: %v\n", err)
+			//I believe the dataSerever will handle the os.Exit(1)
```
Comment (krasi-georgiev) [inline pkg/tracker/runner.go:66]: yeah correct here I don't we should log, but just return the message as an error.

File: pkg/tracker/runner.go [hunk 15]
```diff
@@ -102,9 +105,10 @@ func (r *Runner) Start(ctx context.Context, exitCh chan int) error {
 					//runnerLog.Info("Running trackers...")
```
Comment (krasi-georgiev) [inline pkg/tracker/runner.go:105]: all of these commented logs should be debug logs, but lets leave for another PR.

File: pkg/tracker/runner.go [hunk 16]
```diff
@@ -102,9 +105,10 @@ func (r *Runner) Start(ctx context.Context, exitCh chan int) error {
...
 					go func(count int) {
 						idx := count % len(trackers)
-						err := trackers[idx].Exec(ctx)
+						err := trackers[idx].Exec(ctx, logger)
 						if err != nil {
-							runnerLog.Error("Problem in tracker %s: %v\n", trackers[idx].String(), err)
+							level.Warn(logger).Log("msg", "problem in traker", "tracker", trackers[idx].String(), "err", err)
+							//runnerLog.Error("Problem in tracker %s: %v\n", trackers[idx].String(), err)
```
Comment (krasi-georgiev) [inline pkg/tracker/runner.go:111]:
```suggestion
```

File: pkg/tracker/timeOut.go [hunk 17]
```diff
@@ -46,13 +48,16 @@ func (b *TimeOutTracker) Exec(ctx context.Context) error {
...
+		level.Error(logger).Log("msg", "instance Error, disputeStatus", "err", err)
 		return err
 	}
 	address := "000000000000000000000000" + _fromAddress[2:]
 	decoded, err := hex.DecodeString(address)
 	if err != nil {
-		log.Fatal(err)
+		level.Error(logger).Log("msg", "error decoding address", "err", err)
+		os.Exit(1)
+		//Does log.Fatal terminates the execution automatically?
```
Comment (krasi-georgiev) [inline pkg/tracker/timeOut.go:59]: yes that is why in practise here we shouldn't exit but just pass propagate the error to the caller.

**Review by krasi-georgiev — COMMENTED** [review]
you can now remove the `logConfig_test.go` and also the `config/loggingConfig.json`

There are few other small things and it is ready for merging.

File: cmd/tellor/main.go (hunk 1 above)
Comment (JGcarv) [reply cmd/tellor/main.go:121]: I think it'll be easier to do it in a new one This way we won't add a lot of new changes to this already big Pr!

File: pkg/dataServer/dataServer.go [hunk 18]
```diff
@@ -31,18 +33,21 @@ type DataServer struct {
...
-func CreateServer(ctx context.Context) (*DataServer, error) {
+func CreateServer(ctx context.Context, logger log.Logger) (*DataServer, error) {
 	cfg := config.GetConfig()
 
 	DB := ctx.Value(common.DBContextKey).(db.DB)
 	client := ctx.Value(common.ClientContextKey).(rpc.ETHClient)
 	run, err := tracker.NewRunner(client, DB)
 	if err != nil {
-		log.Fatal(err)
+		level.Error(logger).Log("msg", "erro rcreating tracker runner instance", "err", err)
```
[the remark this answers is not in the document]
Comment (JGcarv) [reply pkg/dataServer/dataServer.go:43]: Question, in this case, the function expects the return `(*DataServer, error)`. Should we return `(nil, errorsNew())`?

File: pkg/ops/disputeOps.go (hunk 9 above)
Comment (JGcarv) [reply pkg/ops/disputeOps.go:175]: The issue here is that the function `simpleCmd` in main expect as a parameter a `func(context.Context, log.Logger) error`, if I change this I'll have to add another function in main to support this. I personally think leaving this here is less impactful than adding another function there. What do you think?

File: pkg/tracker/currentVariables.go [hunk 19]
```diff
@@ -49,43 +47,43 @@ func (b *CurrentVariablesTracker) Exec(ctx context.Context) error {
...
 		bitSetVar = []byte{1}
 	}
-	currentVarsLog.Info("Retrieved variables. challengeHash: %x", currentChallenge)
+	level.Info(logger).Log("msg", "Retrieved variables", "challengeHash", currentChallenge)
 
 	err = DB.Put(db.CurrentChallengeKey, currentChallenge[:])
 	if err != nil {
-		fmt.Println("Current Variables Put Error")
+		level.Error(logger).Log("msg", "Current Variables Put Error", "var", "currentChallenge", "err", err)
 		return err
```
Comment (krasi-georgiev) [inline pkg/tracker/currentVariables.go:62]: I think we should simplify this and the following lines and just return an error here and then the caller can print the log.
```suggestion
		return errors.Wrap(err,"currentChallenge variables put")
```

File: pkg/tracker/disputeChecker.go [hunk 20]
```diff
@@ -91,7 +90,7 @@ func CheckValueAtTime(reqID uint64, val *big.Int, at time.Time) *ValueCheckResul
 	}
 }
 
-func (c *disputeChecker) Exec(ctx context.Context) error {
+func (c *disputeChecker) Exec(ctx context.Context, logger log.Logger) error {
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeChecker.go:93]: Let's add a constructor here so that we can have the component name. the same in all other places for each component.

This
```
func (c *disputeChecker) Exec(ctx context.Context, logger log.Logger){
	logger.
}
```
should change into
```
func NewDisputeChecker(logger log.Logger) *DisputeChecker{
	return &disputeChecker {
		logger:  log.With(logger, "component", "dispute checker")
	}
}

func (c *DisputeChecker) Exec(ctx context.Context){
	c.logger
}
``

File: pkg/tracker/disputeChecker_test.go [hunk 21]
```diff
@@ -65,8 +71,9 @@ func TestDisputeCheckerOutOfRange(t *testing.T) {
 }
 
 func execEthUsdPsrs(ctx context.Context, t *testing.T, psrs []*IndexTracker) {
+	logger := util.SetupLogger("debug")
 	for _, psr := range psrs {
-		err := psr.Exec(ctx)
+		err := psr.Exec(ctx, logger)
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeChecker_test.go:76]: Yeah here we also need to pass the logger to the constructor(so that we can have the component name in the logs) and not the actual struct method.

File: pkg/tracker/gas.go [hunk 22]
```diff
@@ -36,7 +37,7 @@ func (b *GasTracker) String() string {
 	return "GasTracker"
 }
 
-func (b *GasTracker) Exec(ctx context.Context) error {
+func (b *GasTracker) Exec(ctx context.Context, logger log.Logger) error {
```
Comment (krasi-georgiev) [inline pkg/tracker/gas.go:40]: Here we also need a constructor which takes a logger so that can have the component name in the log lines.

The same for all other places.

File: pkg/tracker/newCurrentVariables.go [hunk 23]
```diff
@@ -53,7 +55,7 @@ func (b *NewCurrentVariablesTracker) Exec(ctx context.Context) error {
 	instanceGetter := ctx.Value(tellorCommon.ContractsGetterContextKey).(*getter.TellorGetters)
 	myStatus, err := instanceGetter.DidMine(nil, returnNewVariables.Challenge, fromAddress)
 	if err != nil {
-		fmt.Println("My Status Retrieval Error")
+		level.Error(logger).Log("msg", "My Status Retrieval Error", "err", err)
```
Comment (krasi-georgiev) [inline pkg/tracker/newCurrentVariables.go:58]: For all errors, there is a golang practice to use all lower case and I like to do the same for the logging. Also, I don't put words like `error` in the message as it gets repeated twice in the final output.
```suggestion
		level.Error(logger).Log("msg", "my status retrieval", "err", err)
```

File: pkg/ops/dataServerOps.go [hunk 24]
```diff
@@ -19,31 +20,31 @@ type DataServerOps struct {
...
-func CreateDataServerOps(ctx context.Context, exitCh chan os.Signal) (*DataServerOps, error) {
-	ds, err := dataServer.CreateServer(ctx)
+func CreateDataServerOps(ctx context.Context, logger log.Logger, exitCh chan os.Signal) (*DataServerOps, error) {
+	ds, err := dataServer.CreateServer(ctx, log.With(logger, "component", "ops"))
 	if err != nil {
 		return nil, err
 	}
 	done := make(chan int)
-	ops := &DataServerOps{exitCh: exitCh, server: ds, log: util.NewLogger("ops", "DataServerOps"), done: done, Running: false}
+	ops := &DataServerOps{exitCh: exitCh, server: ds, log: logger, done: done, Running: false}
```
Comment (krasi-georgiev) [inline pkg/ops/dataServerOps.go:34]: Since this is the constructor we want to add the new logger fields here.
```suggestion
	ops := &DataServerOps{exitCh: exitCh, server: ds, log: log.With(logger, "component", "ops"), done: done, Running: false}
```

ping me if I didn't manage to explain properly.

File: pkg/ops/dataServerOps.go [hunk 25]
```diff
@@ -19,31 +20,31 @@ type DataServerOps struct {
...
 
 	return ops, nil
 }
 
 // Start the data server.
 func (ops *DataServerOps) Start(ctx context.Context) error {
-	if err := ops.server.Start(ctx, ops.done); err != nil {
+	if err := ops.server.Start(ctx, ops.log, ops.done); err != nil {
```
Comment (krasi-georgiev) [inline pkg/ops/dataServerOps.go:41]: the logger was already added to the `ops.server` with the `dataServer.CreateServer` constructor.

File: pkg/tracker/balance_test.go [hunk 26]
```diff
@@ -14,10 +14,11 @@ import (
 	"github.com/tellor-io/TellorMiner/pkg/common"
 	"github.com/tellor-io/TellorMiner/pkg/db"
 	"github.com/tellor-io/TellorMiner/pkg/rpc"
+	"github.com/tellor-io/TellorMiner/pkg/util"
 )
 
 func TestStringId(t *testing.T) {
-	tracker := &BalanceTracker{}
+	tracker := &BalanceTracker{logger: util.SetupLogger("debug")}
```
Comment (krasi-georgiev) [inline pkg/tracker/balance_test.go:21]: To be consistent lets use the constructor: `NewBalanceTracker`

File: pkg/tracker/currentVariables.go [hunk 27]
```diff
@@ -5,29 +5,34 @@ package tracker
...
 
 type CurrentVariablesTracker struct {
+	logger log.Logger
 }
 
 func (b *CurrentVariablesTracker) String() string {
 	return CurrentVariablesTrackerName
 }
 
+func New_CurrentVariablesTracker(logger log.Logger) *CurrentVariablesTracker {
```
Comment (krasi-georgiev) [inline pkg/tracker/currentVariables.go:30]:
```suggestion
func NewCurrentVariablesTracker(logger log.Logger) *CurrentVariablesTracker {
```

File: pkg/tracker/currentVariables.go [hunk 28]
```diff
@@ -49,44 +54,37 @@ func (b *CurrentVariablesTracker) Exec(ctx context.Context) error {
...
-		fmt.Println("My Status Retrieval Error")
-		return err
+		return errors.Wrap(err, "Status Retrieval")
 	}
 	bitSetVar := []byte{0}
 	if myStatus {
 		bitSetVar = []byte{1}
 	}
-	currentVarsLog.Info("Retrieved variables. challengeHash: %x", currentChallenge)
+	level.Info(b.logger).Log("msg", "Retrieved variables", "challengeHash", currentChallenge)
```
Comment (krasi-georgiev) [inline pkg/tracker/currentVariables.go:63]: lets be consistent with all lower case for errors and logs.
```suggestion
	level.Info(b.logger).Log("msg", "retrieved variables", "challengeHash", currentChallenge)
```

File: pkg/tracker/disputeChecker_test.go [hunk 29]
```diff
@@ -35,16 +38,17 @@ func TestDisputeCheckerOutOfRange(t *testing.T) {
...
+	disputeChecker := &disputeChecker{lastCheckedBlock: 500, logger: util.SetupLogger("debug")}
 	if _, err := BuildIndexTrackers(); err != nil {
 		t.Fatal(err)
 	}
 	ethUSDPairs := indexes["ETH/USD"]
 	execEthUsdPsrs(ctx, t, ethUSDPairs)
 	time.Sleep(2 * time.Second)
 	execEthUsdPsrs(ctx, t, ethUSDPairs)
-
+	ctx = context.WithValue(ctx, tellorCommon.ContractAddress, common.Address{0x0000000000000000000000000000000000000000})
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeChecker_test.go:49]: Why do we need this? It should be in the CreateContext func

File: pkg/tracker/disputeStatus.go [hunk 30]
```diff
@@ -49,21 +57,21 @@ func (b *DisputeTracker) Exec(ctx context.Context) error {
 
 	instance, err := getter.NewTellorGetters(contractAddress, client)
 	if err != nil {
-		fmt.Println("instance Error, disputeStatus")
+		level.Error(b.logger).Log("msg", "Error getting master instance", "err", err)
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeStatus.go:60]: the `error` word is already added by the logger so no need to have it twice, The same in few other places. Also words like `problem`, `Could not` are also not necessary and when reading the logs are just noise. The logger itself formats errors in a very good an distinct format.
```suggestion
		level.Error(b.logger).Log("msg", "getting master instance", "err", err)
```

File: pkg/tracker/disputeStatus.go [hunk 31]
```diff
@@ -49,21 +57,21 @@ func (b *DisputeTracker) Exec(ctx context.Context) error {
...
 		return err
 	}
 
 	status, _, err := instance.GetStakerInfo(nil, fromAddress)
 
 	if err != nil {
-		fmt.Println("instance Error, disputeStatus")
+		level.Error(b.logger).Log("msg", "Error getting staker info ", "err", err)
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeStatus.go:67]:
```suggestion
		level.Error(b.logger).Log("msg", "getting staker info ", "err", err)
```

File: pkg/tracker/disputeStatus_test.go [hunk 32]
```diff
@@ -14,10 +14,11 @@ import (
 	"github.com/tellor-io/TellorMiner/pkg/common"
 	"github.com/tellor-io/TellorMiner/pkg/db"
 	"github.com/tellor-io/TellorMiner/pkg/rpc"
+	"github.com/tellor-io/TellorMiner/pkg/util"
 )
 
 func TestDisputeString(t *testing.T) {
-	tracker := &DisputeTracker{}
+	tracker := &DisputeTracker{logger: util.SetupLogger("debug")}
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeStatus_test.go:21]: When possible and to be consistent lets use the constructor func to be consistent.

File: pkg/tracker/runner.go [hunk 33]
```diff
@@ -31,16 +31,15 @@ func NewRunner(client rpc.ETHClient, db db.DB) (*Runner, error) {
 }
 
 // Start will kick off the runner until the given exit channel selects.
-func (r *Runner) Start(ctx context.Context, exitCh chan int) error {
+func (r *Runner) Start(ctx context.Context, logger log.Logger, exitCh chan int) error {
```
Comment (krasi-georgiev) [inline pkg/tracker/runner.go:34]: We should add a constructor here as well and add the logger to the instructor like in all the other changes.
```suggestion
func (r *Runner) Start(ctx context.Context, exitCh chan int) error {
```

File: pkg/tracker/timeOut.go (hunk 17 above)
Comment (krasi-georgiev) [reply pkg/tracker/timeOut.go:66]: this still stands, 
would you mind letting me click the resolve button? 
This way it is easier for me to not miss any of the suggestions and changes. 
If you think that some suggestion is incorrect please comment back and will discuss.

File: pkg/tracker/timeOut_test.go [hunk 34]
```diff
@@ -14,10 +14,11 @@ import (
 	"github.com/tellor-io/TellorMiner/pkg/common"
 	"github.com/tellor-io/TellorMiner/pkg/db"
 	"github.com/tellor-io/TellorMiner/pkg/rpc"
+	"github.com/tellor-io/TellorMiner/pkg/util"
 )
 
 func TestTimeOutString(t *testing.T) {
-	tracker := &TimeOutTracker{}
+	tracker := &TimeOutTracker{logger: util.SetupLogger("debug")}
```
Comment (krasi-georgiev) [inline pkg/tracker/timeOut_test.go:21]: use the constructor?

File: pkg/tracker/currentVariables.go (hunk 27 above)
Comment (JGcarv) [reply pkg/tracker/currentVariables.go:30]: This one is a hack because we already have another tracker call `NewCurrentVariables`, so this is the least awkward way of solving that. But I'm open to suggestions!

File: pkg/tracker/timeOut.go (hunk 17 above)
Comment (JGcarv) [reply pkg/tracker/timeOut.go:66]: this one I don't understand what I'm supposed to change? Should I add  `os.Exit(1)` on all thE errors on this file or just return `errors.Wrapf(...)`?

Comment (krasi-georgiev) [reply pkg/tracker/timeOut.go:66]: yes return `errors.Wrap`

Exit should happen only in the main package. All other errors should be either logged(when not that important) or returned to the caller

File: pkg/tracker/currentVariables.go (hunk 27 above)
Comment (krasi-georgiev) [reply pkg/tracker/currentVariables.go:30]: aah yeah this the v1 code. The `pkg/tracker/currentVariables.go` file can be removed and this will resolve the conflict. If you prefer can do it in a parallel PR or can do it even in this one.

File: cmd/tellor/main.go [hunk 35]
```diff
@@ -44,17 +46,19 @@ func buildContext() error {
 		// Create an rpc client
 		client, err := rpc.NewClient(cfg.NodeURL)
 		if err != nil {
-			log.Fatal(err)
+			return fmt.Errorf("Couldn't create client instance: %s", err)
```
Comment (krasi-georgiev) [inline cmd/tellor/main.go:49]: here and few other places, lets use errors.Wrap, all lower case, and no need to use words like - error, couldn't , problem etc. these are obvious from the error log.
```suggestion
			return errors.Wrap(err,"create client instance")
```

File: cmd/tellor/main.go [hunk 36]
```diff
@@ -218,139 +227,149 @@ func newDisputeCmd(cmd *cli.Cmd) {
...
+			ch := make(chan os.Signal)
+			var err error
+			ds, err = ops.CreateDataServerOps(ctx, logger, ch)
+			if err != nil {
+				level.Error(logger).Log("msg", "error creating data server", "err", err)
+				os.Exit(1)
+			}
+			// Start and wait for it to be ready
+			if err := ds.Start(ctx, logger); err != nil {
+				//Should we do this here or pass it down to ExitOnError func for consistency?
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply cmd/tellor/main.go:340]: this one still stands.

File: pkg/tracker/disputeStatus.go [hunk 37]
```diff
@@ -78,15 +86,15 @@ func (b *DisputeTracker) Exec(ctx context.Context) error {
 		//fmt.Println("Getting staker info for address", addr)
 		status, _, err := instance.GetStakerInfo(nil, address)
 		if err != nil {
-			fmt.Printf("Could not get staker dispute status for miner address %s: %v\n", addr, err)
+			level.Error(b.logger).Log("msg", "getting staker dispute status for miner", "address", addr, "err", err)
 		}
 		fmt.Printf("Whitelisted Miner %s Dispute Status: %v\n", addr, status)
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeStatus.go:91]:
```suggestion
```

File: pkg/tracker/factory.go [hunk 38]
```diff
@@ -3,45 +3,49 @@
 
 package tracker
 
-import "fmt"
+import (
+	"fmt"
+
```
Comment (krasi-georgiev) [inline pkg/tracker/factory.go:8]:
```suggestion
```

File: pkg/tracker/gas.go [hunk 39]
```diff
@@ -55,27 +64,27 @@ func (b *GasTracker) Exec(ctx context.Context) error {
 		if err != nil {
 			gasPrice, err = client.SuggestGasPrice(context.Background())
 			if err != nil {
-				log.Println("couldn't get suggested gas price", err)
+				level.Warn(b.logger).Log("msg", "couldn't get suggested gas price", "err", err)
```
Comment (krasi-georgiev) [inline pkg/tracker/gas.go:67]: hm, I think for all trackers we want to return the error to the caller and the caller can decide whether to log it or pass it up the stack. This way we don't even need a logger inside the trackers.
```suggestion
				return errors.Wrap(err, "get suggested gas price")
```

File: pkg/tracker/balance.go [hunk 40]
```diff
@@ -43,10 +50,11 @@ func (b *BalanceTracker) Exec(ctx context.Context) error {
 	balance, err := client.BalanceAt(ctx, fromAddress, nil)
 
 	if err != nil {
-		fmt.Println("balance Error, balance.go")
+		level.Error(b.logger).Log("msg", "getting balance", "err", err)
 		return err
```
Comment (krasi-georgiev) [inline pkg/tracker/balance.go:54]: In places where the functions returns the error and stop executing, don't need to log, just pass the error to the caller and let the caller decide whether to log it or pass it up the stack. This way we can even avoid the need to have a logger in functions that just need to return an error. There are few places like this.

Actually in the trackers I think only the gas tracker needs to log as when it fails to get the price from the gasStation website it can get it from the eth client. In these cases where the execution after an error can continue it is ok to print a log otherwise in functions that stop the execution on every error need to just return the error to the caller and no need to print a log.

This way you can see that `CurrentVariablesTracker`, `TributeTracker` don't even need a logger because they just stop the execution on an error and return the error to the caller and the caller can decide whether to print a log or to pass the error up the stack.

You can leave the constructors though, I am sure we will need a constructor at some point anyway.

```suggestion
		return errors.Wrap(err, "getting balance")
```

File: pkg/tracker/disputeStatus.go [hunk 41]
```diff
@@ -78,15 +86,15 @@ func (b *DisputeTracker) Exec(ctx context.Context) error {
 		//fmt.Println("Getting staker info for address", addr)
```
Comment (krasi-georgiev) [inline pkg/tracker/disputeStatus.go:86]:
```suggestion
```

File: pkg/tracker/timeOut.go [hunk 42]
```diff
@@ -81,14 +88,14 @@ func (b *TimeOutTracker) Exec(ctx context.Context) error {
...
-			log.Fatal(err)
+			return errors.Wrapf(err, "decoding address")
 		}
 		hash := solsha3.SoliditySHA3(decoded)
 		var data [32]byte
 		copy(data[:], hash)
 		status, err := instance.GetUintVar(nil, data)
 		if err != nil {
-			fmt.Printf("Could not get staker timeOut status for miner address %s: %v\n", addr, err)
+			level.Error(b.logger).Log("msg", "getting staker timeOut status for miner", "address", addr, "err", err)
```
Comment (krasi-georgiev) [inline pkg/tracker/timeOut.go:98]: I am not sure about this one. Why would it continue the execution when there is an error?
This means that it will save `0` in the DB. @themandalore can probably give some context here.

File: pkg/tracker/tributeBalance.go [hunk 43]
```diff
@@ -69,9 +77,9 @@ func (b *TributeTracker) Exec(ctx context.Context) error {
 	}
 
 	//numTributes, _ := balanceInTributes.Float64()
-	log.Printf("Tribute Balance: %v (%v tributes)\n", balance, balanceInTributes)
+	fmt.Printf("Tribute Balance: %v (%v tributes)\n", balance, balanceInTributes)
```
Comment (krasi-georgiev) [inline pkg/tracker/tributeBalance.go:80]: convert to a log

File: pkg/tracker/factory.go (hunk 38 above)
Comment (JGcarv) [reply pkg/tracker/factory.go:8]: The extra line is put automatically when I save by the linter... It isn't like this for you?

File: pkg/tracker/gas.go (hunk 39 above)
Comment (JGcarv) [reply pkg/tracker/gas.go:67]: Won't we need the logger in case of Info and Debug messages anyway? 

If not, I don't mind removing the logger from trackers, but can we do it in a fresh PR?

File: pkg/tracker/currentVariables.go (hunk 27 above)
Comment (JGcarv) [reply pkg/tracker/currentVariables.go]: I adjusted the naming but left the tests commented. I'll adjust those in a next PR

File: pkg/tracker/gas.go (hunk 39 above)
Comment (krasi-georgiev) [reply pkg/tracker/gas.go:67]: Yes you are right I just double-checked and some are printing some debug logs.

**Review by krasi-georgiev — COMMENTED** [review]
There are still few places with `fmt.Errorf`, but I opened an issue to do a global search and replace for those https://github.com/tellor-io/TellorMiner/issues/236

**Review by krasi-georgiev — APPROVED** [review]
